genie-agent/src/main/java/com/netflix/genie/agent/execution/services/impl/grpc/GRpcAgentFileStreamServiceImpl.java

```diff
@@ -0,0 +1,363 @@
+import java.nio.channels.FileChannel;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.StandardOpenOption;
+import java.util.ConcurrentModificationException;
+import java.util.NoSuchElementException;
+import java.util.Set;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.Semaphore;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+/**
+ * Implementation of {@link AgentFileStreamService} over gRPC.
+ * Sets up a persistent 2-way stream ('sync') to push manifest updates and receive file requests.
+ * When a file request is received, a creates a new 2 way stream ('transmit') and pushes file chunks, waits for ACK,
+ * sends the next chunk, ... until the file range requested is transmitted. Then the stream is shut down.
+ *
+ * @author mprimi
+ * @since 4.0.0
+ */
+@Slf4j
+public class GRpcAgentFileStreamServiceImpl implements AgentFileStreamService {
+    private static final boolean ENABLE_COMPRESSION = true; //TODO make configurable
+    private static final int MAX_DATA_CHUNK_SIZE = 1024 * 1024; //TODO make configurable
+    private static final int MAX_CONCURRENT_TRANSMIT_STREAMS = 5;  //TODO make configurable
+
+    private final FileStreamServiceGrpc.FileStreamServiceStub fileStreamServiceStub;
+    private final TaskScheduler taskScheduler;
+    private final ExponentialBackOffTrigger trigger;
+    private final JobDirectoryManifestProtoConverter manifestProtoConverter;
+    private final StreamObserver<ServerControlMessage> responseObserver;
+    private final Semaphore concurrentTransfersSemaphore;
+    private final Set<FileTransfer> activeFileTransfers;
+
+    private StreamObserver<AgentManifestMessage> controlStreamObserver;
+    private String jobId;
+    private Path jobDirectoryPath;
+    private AtomicBoolean started = new AtomicBoolean();
+    private ScheduledFuture<?> scheduledTask;
+
+    GRpcAgentFileStreamServiceImpl(
+        final FileStreamServiceGrpc.FileStreamServiceStub fileStreamServiceStub,
+        final TaskScheduler taskScheduler,
+        final JobDirectoryManifestProtoConverter manifestProtoConverter
+    ) {
+        this.fileStreamServiceStub = fileStreamServiceStub;
+        this.taskScheduler = taskScheduler;
+        this.manifestProtoConverter = manifestProtoConverter;
+        this.trigger = new ExponentialBackOffTrigger(
+            ExponentialBackOffTrigger.DelayType.FROM_PREVIOUS_EXECUTION_BEGIN,
+            1000, //TODO make configurable
+            10000, //TODO make configurable
+            1.1f //TODO make configurable
+        );
+        this.responseObserver = new ServerControlStreamObserver(this);
+        this.concurrentTransfersSemaphore = new Semaphore(MAX_CONCURRENT_TRANSMIT_STREAMS);
+        this.activeFileTransfers = Sets.newConcurrentHashSet();
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public synchronized void start(final String claimedJobId, final Path jobDirectoryRoot) {
```

Comment:
This service is a singleton right? Why do we need to synchronize both the start and the stop? Do we think it would get multiple control signals from outside for some reason?

Reply:
Example is a call to `stop()` while running, since there's access to shared variables.